dm_control/locomotion/tasks/reference_pose/rewards.py

```diff
@@ -23,27 +23,27 @@
                                         ['reward', 'debug', 'reward_terms'])
 
 
-def bounded_quat_dist(source, target):
-  """Computes a quaternion distance limiting the difference to a max of pi.
+def bounded_quat_dist(source: np.ndarray,
+                      target: np.ndarray) -> np.ndarray:
+  """Computes a quaternion distance limiting the difference to a max of pi/2.
 
-  This is achieved by taking the minimum of d(source, target) and
-  d(source, -target).
-
-  This function supports batched inputs.
+  This function supports an arbitrary number of batch dimensions, B.
 
   Args:
-    source: a quaternion
-    target: another quaternion
+    source: a quaternion, shape (B, 4).
+    target: another quaternion, shape (B, 4).
 
   Returns:
-    quaternion distance.
+    Quaternion distance, shape (B, 1).
   """
   source /= np.linalg.norm(source, axis=-1, keepdims=True)
   target /= np.linalg.norm(target, axis=-1, keepdims=True)
-  default_dist = tr.quat_dist(source, target)
-  anti_dist = tr.quat_dist(source, -np.asarray(target))
-  min_dist = np.minimum(default_dist, anti_dist)
-  return min_dist
+  dist = 2*np.sum(source * target, axis=-1)**2 - 1
```

Comment:
As I said in my other comment, it's probably faster to write this as `np.einsum('...i,...i', source, target)`.

Also tiny style nit: could you leave spaces around `*` and `**` operators please?

Reply:
I ran a small test, this single line alone runs about 2x faster with `np.einsum`. For the function as a whole, however, there seems to be just a 5-10% increase on top of the previous 10x speedup. I'll add these changes soon, thanks for the suggestions!